feat: Add actions to foreignKey(...)

cegredev opened the pull request:
**Changed:**

Added parameter `actions` to function `foreignKey` in foreign-keys.ts for postgres, mysql and mysql-lite which allows for `onUpdate` and `onDelete` actions when using multi-column foreign keys.

Example:
```ts
const peopleTable = mysqlTable(
	"people",
	{
		firstName: text("first_name"),
		lastName: text("last_name"),
	},
	(table) => ({
		pk: primaryKey({ columns: [table.firstName, table.lastName] }),
	})
);

const homesTable = mysqlTable(
	"homes",
	{
		firstName: text("first_name"),
		lastName: text("last_name"),
		address: text("address"),
	},
	(table) => ({
		fk: foreignKey(
			{
				columns: [table.firstName, table.lastName],
				foreignColumns: [peopleTable.firstName, peopleTable.lastName],
			},
			// THIS IS NEW:
			{
				onDelete: "cascade",
				onUpdate: "cascade",
			}
		),
	})
);
```

**What about tests?**

- In the tests I found, the SQL for each table wasn't generated from the schema but instead written manually
- This means those tests only test the behavior of the non-schema API (the schema you define only serves to provide the types and not for generating the SQL)
- This change is schema API only and I assume the `ForeignKey` class is already well-tested
- Should we write tests for this in drizzle-kit?


**Hi Twitch :3**

cegredev committed:
Add actions to foreignKey(...)

Added parameter actions to function foreignKey in foreign-keys.ts for postgres, mysql and mysql-lite which allows for onUpdate and onDelete actions when using this function instead of references.

AndriiSherman commented:
Thanks for PR!
I guess we can add onUpdate and onDelete to the same config object

```ts
fk: foreignKey({
	columns: [table.firstName, table.lastName],
	foreignColumns: [peopleTable.firstName, peopleTable.lastName],
	onDelete: "cascade", // --> optional 
	onUpdate: "cascade", // --> optional
}),
```

AndriiSherman commented:
Also regarding tests. You can/should test those changes in `drizzle-orm`, by using `getTableConfig`
Please check test [example](https://github.com/drizzle-team/drizzle-orm/blob/main/integration-tests/tests/pg.test.ts#L379)

cegredev committed:
Add actions to config of foreignKey

Removed the `actions` parameter of the `foreignKey` function and instead take `onUpdate` and `onDelete` together with `config`

cegredev committed:
Add tests for postgres and mysql

Added tests using `getTableConfig` for the new `foreignKey(...)` actions feature

cegredev committed:
Fix types and add tests for sqlite

Added types to `foreignKey` for sqlite, not sure if in the right place though. Also added tests.

cegredev committed:
Add more contributing information about tests

Added information on how to run single ava tests and how to test schema-API features

cegredev commented:
Wasn't there live but watched the stream later - thanks for the input!

Made these changes:
- Changed the implementation of `foreignKey` to the way you wanted me to
- Added tests for postgres, mysql and sqlite using `getTableConfig`
    - Wasn't sure where to put sqlite tests, so I put them in `sqlite.proxy.test.ts`, might need to be moved
    - I'm also not testing every possible value of `UpdateDeleteAction` as that seemed overkill
- Added the knowledge I gained from this PR to `CONTRIBUTING.md`

cegredev committed:
Remove accidental auto-format

Accidentally auto-formatted some things that have now been reverted

cegredev committed:
Merge branch 'main' into main

cegredev committed:
Merge branch 'main' into main

cegredev commented:
Is this still missing anything? No worries I get you've got a lot to do, just checking I didn't miss something!

skyf0l commented:
I need this! :tada:

cegredev committed:
Merge branch 'main' into main

sharknoon commented:
Great work cegredev👍 I am very much looking forward to this being merged! This would solve a lot of manual steps in my project. @AndriiSherman Any news regarding your approval? Thank you :)

olup commented:
This feature seems also highly needed on single column foreign key where the autogenerated constraint name is bigger than 64 char - so you want to have a custom naming for the foreign key but also be able to set on delete action ! Can't wait to use it as this is the single blocker for drizzle adoption on our existing database!

Thinking about it, a way to set the fk name in the reference key of the row would also be pretty good, but that's for another PR.

JCraigWasTaken commented:
Bumping this again for activity

AndriiSherman commented:
we will resolve conflicts and release it

Sukairo-02 commented:
Closing this PR due to similar functionality being already implemented:
```typescript
foreignKey({
	columns: [table.firstName, table.lastName],
	foreignColumns: [peopleTable.firstName, peopleTable.lastName],
}).onUpdate('cascade').onDelete('cascade')
```
